File: app/src/main/java/com/example/task2/javafolder/viewviewmodel/fragments/parentfragment/ParentViewModel.java
package com.example.task2.javafolder.viewviewmodel.fragments.parentfragment;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;


import com.example.task2.javafolder.model.Benchmark;
import com.example.task2.javafolder.model.BenchmarksNames;
import com.example.task2.javafolder.model.StartingTimeOfCalculationForBenchmarks;
import com.example.task2.javafolder.model.StateOfCalculation;
import com.example.task2.javafolder.viewviewmodel.fragments.BaseViewModel;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.BackpressureStrategy;
import io.reactivex.Flowable;
import io.reactivex.FlowableOnSubscribe;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import io.reactivex.subscribers.DisposableSubscriber;

public abstract class ParentViewModel extends BaseViewModel {

    private final MutableLiveData<List<Benchmark>> mutableLiveDataForListOfBenchmarks = new MutableLiveData<>();

    private final MutableLiveData<StateOfCalculation> mutableLiveDataForStateOfCalculation = new MutableLiveData<>();

    protected LiveData<List<Benchmark>> liveDataForListOfBenchmarks = mutableLiveDataForListOfBenchmarks;

    protected LiveData<StateOfCalculation> liveDataForStateOfCalculation = mutableLiveDataForStateOfCalculation;

    private Disposable disposable;

    public abstract List<Benchmark> getListOfBenchmarks();

    public abstract long measureTimeOfCalculations(int amountOfItems, BenchmarksNames benchmarksName);

    public void setBenchmarks() {
        mutableLiveDataForListOfBenchmarks.setValue(getListOfBenchmarks());
    }

    public void startCalculations() {
        List<Flowable<Benchmark>> listOfFlowables = createListOfFlowables();
        createDisposableFromFlowables(listOfFlowables);
    }

    public void stopCalculations() {
        stopAnimationOfProgressBars();
        disposeDisposable();
    }

    private List<Flowable<Benchmark>> createListOfFlowables() {
        List<Flowable<Benchmark>> calculationsFlowable = new ArrayList<>();
        for (int i = 0; i < getListOfBenchmarks().size(); i++) {
            final int benchmarksId = i;
            calculationsFlowable.add(Flowable.create((FlowableOnSubscribe<Benchmark>) emitter -> {
                        Benchmark benchmarkBeforeCalculation = createBenchmarkBeforeCalculation(benchmarksId);
                        emitter.onNext(benchmarkBeforeCalculation);
                        Benchmark benchmarkAfterCalculation = createBenchmarkAfterCalculation(benchmarksId);
                        emitter.onNext(benchmarkAfterCalculation);
                        emitter.onComplete();
                    }
                    , BackpressureStrategy.LATEST).subscribeOn(Schedulers.newThread()));
        }
        return calculationsFlowable;
    }

    private Benchmark createBenchmarkBeforeCalculation(int benchmarksId) {
        BenchmarksNames currentBenchmarkName = getListOfBenchmarks().get(benchmarksId).getName();
        return new Benchmark(currentBenchmarkName,
                StartingTimeOfCalculationForBenchmarks.STARTING_TIME.time, StateOfCalculation.CALCULATING);
    }

    private Benchmark createBenchmarkAfterCalculation(int benchmarksId) {
        BenchmarksNames currentBenchmarkName = getListOfBenchmarks().get(benchmarksId).getName();
        long result = measureTimeOfCalculations(repository.getUsersInput(), currentBenchmarkName);
        Benchmark benchmarkAfterCalculation = new Benchmark(currentBenchmarkName, result, StateOfCalculation.TERMINATED);
        benchmarkAfterCalculation.setRepresentationOfCalculationsTime(Long.toString(result));
        return benchmarkAfterCalculation;
    }

    private void createDisposableFromFlowables(List<Flowable<Benchmark>> listOfFlowables) {
        disposeDisposable();
        mutableLiveDataForStateOfCalculation.setValue(StateOfCalculation.CALCULATING);
        if (disposable == null || disposable.isDisposed()) {
            disposable = Flowable.merge(listOfFlowables)
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribeWith(new DisposableSubscriber<Benchmark>() {
                        @Override
                        public void onNext(Benchmark benchmark) {
                            getListOfBenchmarks().set(benchmark.getName().benchmarksId, benchmark);
                            mutableLiveDataForListOfBenchmarks.setValue(getListOfBenchmarks());
                        }

                        @Override
                        public void onError(Throwable t) {
                            disposeDisposable();
                        }

                        @Override
                        public void onComplete() {
                            disposeDisposable();
                        }
                    });
        }
    }

    private void disposeDisposable() {
        mutableLiveDataForStateOfCalculation.setValue(StateOfCalculation.TERMINATED);
        if (disposable != null && !disposable.isDisposed()) {
            disposable.dispose();
        }
    }

    private void stopAnimationOfProgressBars() {
        getListOfBenchmarks().replaceAll(benchmark -> new Benchmark(benchmark.getName(),
                benchmark.getTimeOfCalculations(), StateOfCalculation.TERMINATED));
        mutableLiveDataForListOfBenchmarks.setValue(getListOfBenchmarks());
    }
}

File: app/src/main/java/com/example/task2/kotlinfolder/viewviewmodel/fragments/parentfragment/ParentViewModel.kt
package com.example.task2.kotlinfolder.viewviewmodel.fragments.parentfragment

import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import com.example.task2.kotlinfolder.model.Benchmark
import com.example.task2.kotlinfolder.model.BenchmarksNames
import com.example.task2.kotlinfolder.model.StateOfCalculation
import com.example.task2.kotlinfolder.readOnly
import com.example.task2.kotlinfolder.repo.Repository
import com.example.task2.kotlinfolder.viewviewmodel.fragments.BaseViewModel
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Deferred
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.cancelChildren
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.joinAll
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

abstract class ParentViewModel(repository: Repository) : BaseViewModel(repository) {

    private val mutableLiveDataForListOfBenchmarks = MutableLiveData<List<Benchmark>>()

    val liveDataForListOfBenchmarks = mutableLiveDataForListOfBenchmarks.readOnly()

    private val mutableLiveDataForStateOfCalculation = MutableLiveData<StateOfCalculation>()

    val liveDataForStateOfCalculation = mutableLiveDataForStateOfCalculation.readOnly()

    private var calculationJob: Job? = null

    protected var listOfBenchmarks: MutableList<Benchmark> = mutableListOf()

    private val dispatcherIO: CoroutineDispatcher = Dispatchers.IO

    private val dispatcherMain: CoroutineDispatcher = Dispatchers.Main

    abstract fun fillBenchmarksListWithStartValues()

    abstract fun getCalculationTimeForBenchmark(
        amountOfItems: Int,
        benchmarkName: BenchmarksNames
    ): Long

    override fun onFirstAttach() {
        super.onFirstAttach()
        setBenchmarks()
    }

    private fun setBenchmarks() {
        fillBenchmarksListWithStartValues()
        mutableLiveDataForListOfBenchmarks.value = listOfBenchmarks
    }

    fun startCalculations() {
        mutableLiveDataForStateOfCalculation.value = StateOfCalculation.CALCULATING
        launchCoroutinesFromFlows()
    }

    fun stopCalculations() {
        mutableLiveDataForStateOfCalculation.value = StateOfCalculation.TERMINATED
        calculationJob?.cancelChildren()
        stopAnimationOfProgressBars()
    }


    private fun createBenchmarkBeforeCalculation(benchmarksId: Int): Benchmark {
        val currentBenchmarkName = listOfBenchmarks[benchmarksId].name
        return Benchmark(
            currentBenchmarkName,
            null,
            StateOfCalculation.CALCULATING
        )
    }

    private fun createBenchmarkAfterCalculation(benchmarksId: Int): Benchmark {
        val currentBenchmarkName = listOfBenchmarks[benchmarksId].name
        val result = getCalculationTimeForBenchmark(repository.usersInput, currentBenchmarkName)
        return Benchmark(
            currentBenchmarkName,
            result,
            StateOfCalculation.TERMINATED,
            result.toString()
        )
    }


    private fun launchCoroutinesFromFlows() {
        val jobs = mutableListOf<Job>()
        viewModelScope.launch(dispatcherIO) {
            for (i in 0 until listOfBenchmarks.size) {
                val job = launch {
                    listOfBenchmarks[i] = createBenchmarkBeforeCalculation(i)
                    withContext(dispatcherMain) {
                        mutableLiveDataForListOfBenchmarks.value = listOfBenchmarks
                    }
                    listOfBenchmarks[i] = createBenchmarkAfterCalculation(i)
                    withContext(dispatcherMain) {
                        mutableLiveDataForListOfBenchmarks.value = listOfBenchmarks
                    }
                }
                jobs.add(job)
            }
            jobs.joinAll()

            withContext(dispatcherMain) {
                mutableLiveDataForStateOfCalculation.value = StateOfCalculation.TERMINATED
            }
        }
    }


    private fun stopAnimationOfProgressBars() {
        listOfBenchmarks.replaceAll { benchmark ->
            Benchmark(
                benchmark.name,
                benchmark.timeOfCalculation,
                StateOfCalculation.TERMINATED,
                benchmark.timeOfCalculationToString
            )
        }
        mutableLiveDataForListOfBenchmarks.value = listOfBenchmarks
    }
}
